### ISSUE TITLE
Require current password before accepting a new password

### ISSUE THREAD

**Issue by b-jazz** [openstreetmap/openstreetmap-website#2144]
If I find a computer in a lab with a browser that is logged into osm.org, I can change the password without knowing the current password and take over the person's account. Most other password reset flows will require knowing the current password before changing to a new password.

**Comment by tomhughes** [comment]
If people leave a logged in account where other people can use it then it's pretty much game over anyway so I don't see this as a major issue. I won't object if somebody want to change it, but it will be hard to do as things stand.

You would also have to stop people changing the social login provider, or adding a new one, or adding a password to an account that didn't have one and that would be even harder as you'd have to revalidate the social login (and hope they hadn't also left google/facebook/whatever logged in as well).

**Comment by pnorman** [comment]
I agree with changing it, but also that there's a lot of cases to cover with social logins, etc, all of which require re-authentication.

**Comment by mmd-osm** [comment]
Being able to change your email address without entering a password is probably also something that needs to be revisited here.

**Comment by tomhughes** [comment]
Not really because changing that requires that you demonstrate control of the current email by being able to read an email sent to it.

**Comment by mmd-osm** [comment]
My comment was really  meant for the use case „If I find a computer in a lab with a browser that is logged into osm.org“...

**Comment by tomhughes** [comment]
Yes and my point is that just being logged in to osm.org doesn't let you change the email - it will be recorded as a pending change and a link sent to the old email that has to be clicked to confirm the change.

**Comment by mmd-osm** [comment]
Are you sure? I tested this an hour ago and the confirmation email was sent to the new email address (=old address with a different + suffix)

**Comment by tomhughes** [comment]
Ah yes you might be right, so yes that's just another reason why it's game over if you leave a machine logged in.

Sure we should probably improve it, but because the authentication and settings stuff is a complete clusterfsck it will be a massive amount of work and should probably only be tackled once that is sorted or we'll just break everything.

**Comment by gravitystorm** [comment]
> Sure we should probably improve it, but because the authentication and settings stuff is a complete clusterfsck it will be a massive amount of work and should probably only be tackled once that is sorted or we'll just break everything.

Yes, I think the best approach would be to move to Devise, and then work on these problems there. For example, in another project we recently required changing emails to a) send a warning to the existing email address and b) also send confirmation link to the new email address. These are just configuration settings in devise, we didn't need to write code for any of that.

**Comment by mmd-osm** [comment]
Some sites ask a user to go through a "forget password" first in case they previously signed up using a third party auth provider and lack a dedicated password for the site.

That forget password function would then send an email to the user, asking them to set a new password. All other changes in the settings dialog should always ask for that current password, irrespective of the login method in use (third party social or password). The good thing is that we already have a forget password function, and it would probably solve a lot of hassle with any kind of third party providers - the logon mechanism simply becomes irrelevant when it comes to changing your settings.

Example from coverity.com:

![password_forget](https://user-images.githubusercontent.com/5842757/52902456-3f56d980-3211-11e9-8e57-95d1b8072ea3.png)

Re changing the password: I wonder if it would sense to de-authenticate all OAuth clients as well, as they would otherwise still have access in the user's name. Maybe some kind of recommendation text to manually remove OAuth clients would be a good first step, though.
